File: server/infra/cache.py
from redis.asyncio import Redis, RedisCluster

from infra.config import RedisConfig
from infra.log import logger

_redis: Redis | RedisCluster | None = None


async def init_redis(config: RedisConfig) -> Redis | RedisCluster:
    """Create and return a Redis client (standalone or cluster)."""
    global _redis

    if config.cluster:
        _redis = RedisCluster(
            host=config.host,
            port=config.port,
            password=config.password or None,
            decode_responses=True,
        )
    else:
        _redis = Redis(
            host=config.host,
            port=config.port,
            password=config.password or None,
            db=config.db,
            decode_responses=True,
        )

    await _redis.ping()
    mode = "cluster" if config.cluster else "standalone"
    logger.info("Redis connected ({}): {}:{}", mode, config.host, config.port)
    return _redis


def get_redis() -> Redis | RedisCluster:
    """Return the global Redis client. Must call init_redis() first."""
    if _redis is None:
        raise RuntimeError("Redis not initialized. Call init_redis() first.")
    return _redis


async def close_redis() -> None:
    """Close the Redis connection."""
    global _redis
    if _redis is not None:
        await _redis.aclose()
        logger.info("Redis connection closed")
    _redis = None

File: backend/internal/infra/cache.go
package infra

import (
	"context"
	"fmt"
	"strings"
	"time"

	"github.com/redis/go-redis/v9"
	"github.com/rs/zerolog/log"

	"astron-claw/backend/internal/config"
)

func InitRedis(cfg config.RedisConfig) (redis.UniversalClient, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	var rdb redis.UniversalClient

	if cfg.IsCluster() {
		rdb = redis.NewClusterClient(&redis.ClusterOptions{
			Addrs:        cfg.Addrs,
			Password:     cfg.Password,
			PoolSize:     cfg.PoolSize,
			MinIdleConns: cfg.MinIdleConns,
		})
	} else {
		rdb = redis.NewClient(&redis.Options{
			Addr:         cfg.Addrs[0],
			Password:     cfg.Password,
			DB:           cfg.DB,
			PoolSize:     cfg.PoolSize,
			MinIdleConns: cfg.MinIdleConns,
		})
	}

	if err := rdb.Ping(ctx).Err(); err != nil {
		return nil, fmt.Errorf("redis ping: %w", err)
	}

	mode := "standalone"
	addr := cfg.Addrs[0]
	if cfg.IsCluster() {
		mode = "cluster"
		addr = strings.Join(cfg.Addrs, ",")
	}

	log.Info().
		Str("mode", mode).
		Str("addr", addr).
		Msg("Redis connected")

	return rdb, nil
}

func CloseRedis(rdb redis.UniversalClient) {
	if rdb == nil {
		return
	}
	if err := rdb.Close(); err != nil {
		log.Error().Err(err).Msg("Failed to close Redis")
	} else {
		log.Info().Msg("Redis connection closed")
	}
}
